### PR TITLE
exec: outer join

### PR DESCRIPTION
**Pull request by changangela**
This PR adds right and left outer join support. It builds off of #33237 (template hash joiner).

```
BenchmarkHashJoiner/nulls=false/fullOuter=false/distinct=true/rows=2048-8         	   10000	    190102 ns/op	 689.48 MB/s
BenchmarkHashJoiner/nulls=false/fullOuter=false/distinct=true/rows=262144-8       	     100	  12164518 ns/op	1379.19 MB/s
BenchmarkHashJoiner/nulls=false/fullOuter=false/distinct=true/rows=4194304-8      	      10	 176378822 ns/op	1521.93 MB/s
BenchmarkHashJoiner/nulls=false/fullOuter=false/distinct=false/rows=2048-8        	   10000	    208850 ns/op	 627.59 MB/s
BenchmarkHashJoiner/nulls=false/fullOuter=false/distinct=false/rows=262144-8      	     100	  18943393 ns/op	 885.65 MB/s
BenchmarkHashJoiner/nulls=false/fullOuter=false/distinct=false/rows=4194304-8     	       2	 666980843 ns/op	 402.46 MB/s
BenchmarkHashJoiner/nulls=false/fullOuter=true/distinct=true/rows=2048-8          	    5000	    275679 ns/op	 475.45 MB/s
BenchmarkHashJoiner/nulls=false/fullOuter=true/distinct=true/rows=262144-8        	     100	  16794993 ns/op	 998.94 MB/s
BenchmarkHashJoiner/nulls=false/fullOuter=true/distinct=true/rows=4194304-8       	       5	 278006221 ns/op	 965.57 MB/s
BenchmarkHashJoiner/nulls=false/fullOuter=true/distinct=false/rows=2048-8         	    5000	    347624 ns/op	 377.05 MB/s
BenchmarkHashJoiner/nulls=false/fullOuter=true/distinct=false/rows=262144-8       	      50	  26753077 ns/op	 627.11 MB/s
BenchmarkHashJoiner/nulls=false/fullOuter=true/distinct=false/rows=4194304-8      	       2	 510258632 ns/op	 526.08 MB/s
BenchmarkHashJoiner/nulls=true/fullOuter=false/distinct=true/rows=2048-8          	    3000	    384740 ns/op	 340.68 MB/s
BenchmarkHashJoiner/nulls=true/fullOuter=false/distinct=true/rows=262144-8        	      50	  33669674 ns/op	 498.29 MB/s
BenchmarkHashJoiner/nulls=true/fullOuter=false/distinct=true/rows=4194304-8       	       2	 528106688 ns/op	 508.30 MB/s
BenchmarkHashJoiner/nulls=true/fullOuter=false/distinct=false/rows=2048-8         	    3000	    437200 ns/op	 299.80 MB/s
BenchmarkHashJoiner/nulls=true/fullOuter=false/distinct=false/rows=262144-8       	      30	  41442007 ns/op	 404.84 MB/s
BenchmarkHashJoiner/nulls=true/fullOuter=false/distinct=false/rows=4194304-8      	       2	 970349145 ns/op	 276.64 MB/s
BenchmarkHashJoiner/nulls=true/fullOuter=true/distinct=true/rows=2048-8           	    5000	    337161 ns/op	 388.75 MB/s
BenchmarkHashJoiner/nulls=true/fullOuter=true/distinct=true/rows=262144-8         	      50	  28320695 ns/op	 592.40 MB/s
BenchmarkHashJoiner/nulls=true/fullOuter=true/distinct=true/rows=4194304-8        	       3	 431805131 ns/op	 621.66 MB/s
BenchmarkHashJoiner/nulls=true/fullOuter=true/distinct=false/rows=2048-8          	    5000	    344703 ns/op	 380.25 MB/s
BenchmarkHashJoiner/nulls=true/fullOuter=true/distinct=false/rows=262144-8        	      50	  34031242 ns/op	 492.99 MB/s
BenchmarkHashJoiner/nulls=true/fullOuter=true/distinct=false/rows=4194304-8       	       2	 848903163 ns/op	 316.21 MB/s
```

### CODE REVIEW

**Review by asubiotto — COMMENTED** [review]
Haven't reviewed this fully yet, as I first need to review the `hashJoiner` templating PR, but I'm wondering if there are hash join logic tests you can add the `local-vec` config to (I see you added it to `select` and `interleaved`)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33206#-:-LU5N9J1AWFOqLF6whsN:br2zdp9)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/exec/hashjoiner_test.go, line 79 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/33206#-LU5Mqpq4EblpmZd-ZHc:-LU5Mqpq4EblpmZd-ZHd:b-hr2jrp) ([raw file](https://github.com/cockroachdb/cockroach/blob/2acbe1ac48b2cbf24aeb569f01b6a08e7aad3d96/pkg/sql/exec/hashjoiner_test.go#L79)):*
> ```Go
> 			rightOutCols: []uint32{0},
> 
> 			rightOuter:    true,
> ```

Consider using `sqlbase.JoinType`

**Review by changangela — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33206)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/exec/hashjoiner_test.go, line 79 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/33206#-LU5Mqpq4EblpmZd-ZHc:-LU70mkR-kZsgxYnxwzO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2acbe1ac48b2cbf24aeb569f01b6a08e7aad3d96/pkg/sql/exec/hashjoiner_test.go#L79)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Consider using `sqlbase.JoinType`
</blockquote></details>

Done.

**Review by asubiotto — CHANGES_REQUESTED** [review]
Did a first pass. I wonder if you could also take a look at the `distsqlrun/hashjoiner_test.go` unit tests, they might inspire you to test some edge cases you might have not thought about.

@jordanlewis, could you take a quick look as well?  I'm not too familiar with the hash joiner code.

> Reviewed 2 of 8 files at r11.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33206#-:-LUAOllV6MtxkJRzsSVC:bjb38z1)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/distsqlrun/column_exec_setup.go, line 163 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/33206#-LUA3hv_6bqxQ5GA0YSI:-LUA3hva81sM8b8FsWEy:b52j754) ([raw file](https://github.com/cockroachdb/cockroach/blob/c68b27b467255bb5e12b9346f3a67fa8f27abfe5/pkg/sql/distsqlrun/column_exec_setup.go#L163)):*
> ```Go
> 
> 		var leftOuter, rightOuter bool
> 		switch core.HashJoiner.Type {
> ```

Rather than have a switch statement that sets booleans, Can you not have the `hashJoiner` take in a `sqlbase.JoinType`?.
___
*[pkg/sql/exec/colvec.go, line 80 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/33206#-LUAAqKg059pgMJJvUWj:-LUAAqKg059pgMJJvUWk:bgx8xrp) ([raw file](https://github.com/cockroachdb/cockroach/blob/c68b27b467255bb5e12b9346f3a67fa8f27abfe5/pkg/sql/exec/colvec.go#L80)):*
> ```Go
> 	// CopyWithSelAndNilsInt64 copies vec, filtered by sel, unless nils is set,
> 	// into ColVec. It replaces the contents of this ColVec.
> 	CopyWithSelAndNilsInt64(vec ColVec, sel []uint64, nSel uint16, nils []bool, colType types.T)
> ```

It's kind of weighing on me that we now have 4 variants of `Copy`. Not sure what to do about it and don't think we should do anything about it in this PR but we might have to think about this design cc @jordanlewis.
___
*[pkg/sql/exec/colvec.go, line 183 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/33206#-LUAAY2M3IpyQqslUgEe:-LUAAY2M3IpyQqslUgEf:b-qowia5) ([raw file](https://github.com/cockroachdb/cockroach/blob/c68b27b467255bb5e12b9346f3a67fa8f27abfe5/pkg/sql/exec/colvec.go#L183)):*
> ```Go
> 	m.hasNulls = true
> 	for i := range m.nulls {
> 		m.nulls[i] = ^0
> ```

You might be able to reuse the same strategy as in `UnsetNulls` with the opposite of `zeroedNulls` which is initialized with an `init` function
___
*[pkg/sql/exec/hashjoiner.go, line 99 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/33206#-LUANH6L4kUxrnaZ6RrL:-LUANH6L4kUxrnaZ6RrM:boj9zc5) ([raw file](https://github.com/cockroachdb/cockroach/blob/c68b27b467255bb5e12b9346f3a67fa8f27abfe5/pkg/sql/exec/hashjoiner.go#L99)):*
> ```Go
> // Depending on the value of the buildDistinct flag, there are two variations of
> // the probe phase. The planner will set buildDistinct to true if and only if
> // either the left equality columns or the right equality columns make a
> ```

Does this comment need to be updated at all?
___
*[pkg/sql/exec/hashjoiner.go, line 154 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/33206#-LUAFD0LDWX8FTM4EA0y:-LUAFD0LDWX8FTM4EA0z:bd52d6b) ([raw file](https://github.com/cockroachdb/cockroach/blob/c68b27b467255bb5e12b9346f3a67fa8f27abfe5/pkg/sql/exec/hashjoiner.go#L154)):*
> ```Go
> //    to ensure that each batch is at most ColBatchSize.
> 
> type hashJoinEqInnerOp struct {
> ```

Since this now does outer joins, this can now be renamed to `hashJoinerOp` or something like that.
___
*[pkg/sql/exec/hashjoiner.go, line 268 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/33206#-LUAFrXw5I3b0TEmmww4:-LUAFrXw5I3b0TEmmww5:b-m4htyq) ([raw file](https://github.com/cockroachdb/cockroach/blob/c68b27b467255bb5e12b9346f3a67fa8f27abfe5/pkg/sql/exec/hashjoiner.go#L268)):*
> ```Go
> 
> func (hj *hashJoinEqInnerOp) emitUnmatched() {
> 	nResults := uint16(0)
> ```

Could there be overflow here?
___
*[pkg/sql/exec/hashjoiner.go, line 569 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/33206#-LUAM1dR3lZGAZbqlsdx:-LUAM1dR3lZGAZbqlsdy:b-izf3ow) ([raw file](https://github.com/cockroachdb/cockroach/blob/c68b27b467255bb5e12b9346f3a67fa8f27abfe5/pkg/sql/exec/hashjoiner.go#L569)):*
> ```Go
> 
> 	buildIdx []uint64
> 	probeIdx []uint16
> ```

OOC, why is `buildIdx` a `[]uint64` and `probeIdx` a `[]uint16`? Might be nice to add a comment to both of these.
___
*[pkg/sql/exec/hashjoiner.go, line 575 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/33206#-LUAMH_k24kFqz4BfC6a:-LUAMH_k24kFqz4BfC6b:bmeldc4) ([raw file](https://github.com/cockroachdb/cockroach/blob/c68b27b467255bb5e12b9346f3a67fa8f27abfe5/pkg/sql/exec/hashjoiner.go#L575)):*
> ```Go
> 
> 	buildColOffset uint32
> 	probeColOffset uint32
> ```

Consider adding comments to all of these fields.
___
*[pkg/sql/exec/hashjoiner.go, line 617 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/33206#-LUAMa-i9rdbGKE2DPiH:-LUAMa-i9rdbGKE2DPiI:bf5cqla) ([raw file](https://github.com/cockroachdb/cockroach/blob/c68b27b467255bb5e12b9346f3a67fa8f27abfe5/pkg/sql/exec/hashjoiner.go#L617)):*
> ```Go
> ) *hashJoinProber {
> 	// Prepare the output batch by allocating with the correct column types.
> ```

nit: extra line
___
*[pkg/sql/exec/hashjoiner.go, line 906 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/33206#-LUAE08w6u4yNjM7_VNr:-LUAE08w6u4yNjM7_VNs:bnd1l1u) ([raw file](https://github.com/cockroachdb/cockroach/blob/c68b27b467255bb5e12b9346f3a67fa8f27abfe5/pkg/sql/exec/hashjoiner.go#L906)):*
> ```Go
> 			sourceTypes: leftTypes,
> 			source:      leftSource,
> 			outer:       leftOuter,
> ```

It feels a bit strange to me to set these booleans in the source spec. I would set the `sqlbase.JoinType` in the `hashJoinerSpec` and have the phases that need to refer to that. What do you think about that? Maybe there isn't a good reason to do so though.

**Review by changangela — COMMENTED** [review]
Thanks for the review! I've addressed the comments and working on improving logic tests.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33206#-:-LUBUHUWA1YMIx4cuXdT:b-4a9u2x)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/distsqlrun/column_exec_setup.go, line 163 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/33206#-LUA3hv_6bqxQ5GA0YSI:-LUBNeNH29pYwyFRDYbQ:b-qd5com) ([raw file](https://github.com/cockroachdb/cockroach/blob/c68b27b467255bb5e12b9346f3a67fa8f27abfe5/pkg/sql/distsqlrun/column_exec_setup.go#L163)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Rather than have a switch statement that sets booleans, Can you not have the `hashJoiner` take in a `sqlbase.JoinType`?.
</blockquote></details>

The reason I wanted to give each source an `outer` is because this allows us to not worry about left and right, and just dynamically choose the build/probe tables at runtime. If we use `sqlbase.JoinType` instead, we would have to have special logic to worry about if its a LeftJoin or FullJoin in the case that the left table is the build table, etc. I think that translating the `sqlbase.JoinType` to these `outer` flags during planning makes the logic a lot simpler in the joiner.  Alternatively, we can propagate this flag further down but I'm not sure how useful that is with this approach. What do you think?
___
*[pkg/sql/exec/colvec.go, line 80 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/33206#-LUAAqKg059pgMJJvUWj:-LUBOWhK3uqxJ3IW3PwN:b-7qeqxd) ([raw file](https://github.com/cockroachdb/cockroach/blob/c68b27b467255bb5e12b9346f3a67fa8f27abfe5/pkg/sql/exec/colvec.go#L80)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

It's kind of weighing on me that we now have 4 variants of `Copy`. Not sure what to do about it and don't think we should do anything about it in this PR but we might have to think about this design cc @jordanlewis.
</blockquote></details>

Done. Removed one of them :')  I think we also mentioned at some point that we might want to separate out `memColumn` to `memColumn16 ` and `memColumn64`. (I'm not sure how the `ColVec` interface would work though)
___
*[pkg/sql/exec/colvec.go, line 183 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/33206#-LUAAY2M3IpyQqslUgEe:-LUBOq9q5yBnQCnPKYRy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c68b27b467255bb5e12b9346f3a67fa8f27abfe5/pkg/sql/exec/colvec.go#L183)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

You might be able to reuse the same strategy as in `UnsetNulls` with the opposite of `zeroedNulls` which is initialized with an `init` function
</blockquote></details>

Done.
___
*[pkg/sql/exec/hashjoiner.go, line 99 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/33206#-LUANH6L4kUxrnaZ6RrL:-LUBRwF62RiBjOLyjf8M:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c68b27b467255bb5e12b9346f3a67fa8f27abfe5/pkg/sql/exec/hashjoiner.go#L99)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Does this comment need to be updated at all?
</blockquote></details>

Done.
___
*[pkg/sql/exec/hashjoiner.go, line 154 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/33206#-LUAFD0LDWX8FTM4EA0y:-LUBRv_x8pG6kiyMnMxn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c68b27b467255bb5e12b9346f3a67fa8f27abfe5/pkg/sql/exec/hashjoiner.go#L154)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Since this now does outer joins, this can now be renamed to `hashJoinerOp` or something like that.
</blockquote></details>

Done.
___
*[pkg/sql/exec/hashjoiner.go, line 268 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/33206#-LUAFrXw5I3b0TEmmww4:-LUBS1cd2uFnYeKTnqnu:bt7834z) ([raw file](https://github.com/cockroachdb/cockroach/blob/c68b27b467255bb5e12b9346f3a67fa8f27abfe5/pkg/sql/exec/hashjoiner.go#L268)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Could there be overflow here?
</blockquote></details>

We should be okay as long as `ColBatchSize` fits in `uint16` I think.
___
*[pkg/sql/exec/hashjoiner.go, line 569 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/33206#-LUAM1dR3lZGAZbqlsdx:-LUBSe1N5Nq83wfuWO4N:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c68b27b467255bb5e12b9346f3a67fa8f27abfe5/pkg/sql/exec/hashjoiner.go#L569)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

OOC, why is `buildIdx` a `[]uint64` and `probeIdx` a `[]uint16`? Might be nice to add a comment to both of these.
</blockquote></details>

Done.
___
*[pkg/sql/exec/hashjoiner.go, line 575 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/33206#-LUAMH_k24kFqz4BfC6a:-LUBTI-22CN1q0fTzmxC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c68b27b467255bb5e12b9346f3a67fa8f27abfe5/pkg/sql/exec/hashjoiner.go#L575)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Consider adding comments to all of these fields.
</blockquote></details>

Done.
___
*[pkg/sql/exec/hashjoiner.go, line 906 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/33206#-LUAE08w6u4yNjM7_VNr:-LUBTxsC93SIgnngZAwn:bjmecu4) ([raw file](https://github.com/cockroachdb/cockroach/blob/c68b27b467255bb5e12b9346f3a67fa8f27abfe5/pkg/sql/exec/hashjoiner.go#L906)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

It feels a bit strange to me to set these booleans in the source spec. I would set the `sqlbase.JoinType` in the `hashJoinerSpec` and have the phases that need to refer to that. What do you think about that? Maybe there isn't a good reason to do so though.
</blockquote></details>

Hmm I've added a comment about this somewhere above. Maybe I can propagate the `sqlbase.JoinType` into this function and do a switch statement here?

**Review by asubiotto — APPROVED** [review]
LGTM from me but I would wait for Jordan's approval

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33206#-:-LUBlPTF15vodP-m63ER:b-im0al1)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/distsqlrun/column_exec_setup.go, line 163 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/33206#-LUA3hv_6bqxQ5GA0YSI:-LUBkLqzDtKIVhGrRFNx:bwduvs4) ([raw file](https://github.com/cockroachdb/cockroach/blob/c68b27b467255bb5e12b9346f3a67fa8f27abfe5/pkg/sql/distsqlrun/column_exec_setup.go#L163)):*
<details><summary><i>Previously, changangela (Angela Chang) wrote…</i></summary><blockquote>

The reason I wanted to give each source an `outer` is because this allows us to not worry about left and right, and just dynamically choose the build/probe tables at runtime. If we use `sqlbase.JoinType` instead, we would have to have special logic to worry about if its a LeftJoin or FullJoin in the case that the left table is the build table, etc. I think that translating the `sqlbase.JoinType` to these `outer` flags during planning makes the logic a lot simpler in the joiner.  Alternatively, we can propagate this flag further down but I'm not sure how useful that is with this approach. What do you think?
</blockquote></details>

As I state below, I think we're going to have to do this in the future anyway, although I'm fine with this for now. Additionally, I think that in general, execution should not concern itself with trying to determine which side to use as the build/probe side and should expect that the optimized plan has resulted in the right side being used for building. This is definitely not true now though but would, for example make the logic simpler.
___
*[pkg/sql/exec/hashjoiner.go, line 906 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/33206#-LUAE08w6u4yNjM7_VNr:-LUBj0LVDTINHDwnQzmP:bxdrk63) ([raw file](https://github.com/cockroachdb/cockroach/blob/c68b27b467255bb5e12b9346f3a67fa8f27abfe5/pkg/sql/exec/hashjoiner.go#L906)):*
<details><summary><i>Previously, changangela (Angela Chang) wrote…</i></summary><blockquote>

Hmm I've added a comment about this somewhere above. Maybe I can propagate the `sqlbase.JoinType` into this function and do a switch statement here?
</blockquote></details>

I think that might be a good in-between. I believe we might want to support other types of joins that might require more logic in the `hashJoiner`, so it's information we're going to have to pipe down at some point.

**Review by changangela — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33206)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/distsqlrun/column_exec_setup.go, line 163 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/33206#-LUA3hv_6bqxQ5GA0YSI:-LUBz1XW7Ykq7F0TihqG:b-y9mz6i) ([raw file](https://github.com/cockroachdb/cockroach/blob/c68b27b467255bb5e12b9346f3a67fa8f27abfe5/pkg/sql/distsqlrun/column_exec_setup.go#L163)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

As I state below, I think we're going to have to do this in the future anyway, although I'm fine with this for now. Additionally, I think that in general, execution should not concern itself with trying to determine which side to use as the build/probe side and should expect that the optimized plan has resulted in the right side being used for building. This is definitely not true now though but would, for example make the logic simpler.
</blockquote></details>

Done. Good point
___
*[pkg/sql/exec/hashjoiner.go, line 906 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/33206#-LUAE08w6u4yNjM7_VNr:-LUBz-uI2Y5mO31Y6GMt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c68b27b467255bb5e12b9346f3a67fa8f27abfe5/pkg/sql/exec/hashjoiner.go#L906)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I think that might be a good in-between. I believe we might want to support other types of joins that might require more logic in the `hashJoiner`, so it's information we're going to have to pipe down at some point.
</blockquote></details>

Done.

**Review by jordanlewis — APPROVED** [review]
:lgtm: great work!

> Reviewed 1 of 14 files at r12, 1 of 1 files at r13, 5 of 8 files at r14, 3 of 3 files at r15, 1 of 1 files at r16.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33206#-:-LUC5b5Z1utUzJwwm3AD:b-2zonaq)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/sql/exec/colvec.go, line 80 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/33206#-LUAAqKg059pgMJJvUWj:-LUC3b-F1yNOL2-kF_xp:bwb2ygl) ([raw file](https://github.com/cockroachdb/cockroach/blob/c68b27b467255bb5e12b9346f3a67fa8f27abfe5/pkg/sql/exec/colvec.go#L80)):*
<details><summary><i>Previously, changangela (Angela Chang) wrote…</i></summary><blockquote>

Done. Removed one of them :')  I think we also mentioned at some point that we might want to separate out `memColumn` to `memColumn16 ` and `memColumn64`. (I'm not sure how the `ColVec` interface would work though)
</blockquote></details>

Yeah agreed that this is getting messy and gross. Let's think of a way to fix it soon now that there are lots of examples and use cases.
___
*[pkg/sql/exec/colvec_tmpl.go, line 178 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/33206#-LUC3uz37drg9bXHq93a:-LUC3uz37drg9bXHq93b:b-8ao16p) ([raw file](https://github.com/cockroachdb/cockroach/blob/49bced034a70e5a28622a775c0825d1655aeb0df/pkg/sql/exec/colvec_tmpl.go#L178)):*
> ```Go
> 			for i := uint16(0); i < nSel; i++ {
> 				if nils[i] {
> 					m.SetNull(i)
> ```

This whole loop is very sketchy - I see what you mean now about the overuse of SetNull and NullAt. We need to think of a better plan here. Can you add `TODO(jordan): copy the null arrays in batch.`
___
*[pkg/sql/exec/hashjoiner.go, line 687 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/33206#-LUC54NrADIjXSdGnlCS:-LUC54NsCE9WHqqcSotI:b-13bb2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/49bced034a70e5a28622a775c0825d1655aeb0df/pkg/sql/exec/hashjoiner.go#L687)):*
> ```Go
> }
> 
> // exec a general prober that works with non-distinct build table equality
> ```

exec is*

**Review by changangela — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33206)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/exec/colvec_tmpl.go, line 178 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/33206#-LUC3uz37drg9bXHq93a:-LUC6SCu57b_1FaLcRun:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/49bced034a70e5a28622a775c0825d1655aeb0df/pkg/sql/exec/colvec_tmpl.go#L178)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

This whole loop is very sketchy - I see what you mean now about the overuse of SetNull and NullAt. We need to think of a better plan here. Can you add `TODO(jordan): copy the null arrays in batch.`
</blockquote></details>

Done.
___
*[pkg/sql/exec/hashjoiner.go, line 687 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/33206#-LUC54NrADIjXSdGnlCS:-LUC6ikn9nc2rto9WhaQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/49bced034a70e5a28622a775c0825d1655aeb0df/pkg/sql/exec/hashjoiner.go#L687)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

exec is*
</blockquote></details>

Done.

**Pull request merged**
